### PR TITLE
Add support for no-arg commands in Dockerfile

### PR DESCRIPTION
**Pull request by duglin**
We're hoping to add some new commands that don't have any args so this
PR will enable that by removing all of the hard-coded checks that require
commands to have at least one arg.  It also adds some checks to each
command so we're consistent in the error message we get.  Added a test
for this too.

We actually had this check in at least 3 different places (twice in the
parser and once in most cmds), this removes 2 of them (the parser ones).

As part of this I used to have code to throw an error when there was a
Dockerfile command that was unknown.  However, I realized that while I
spotted it during this work, it really is a separate issue and should be discussed
on its own so I'm moving that into a new PR ( https://github.com/docker/docker/pull/10591 ).

Signed-off-by: Doug Davis dug@us.ibm.com

### CODE REVIEW

**Comment by tianon** [comment]
:tada: this was bothering me, but I didn't see any reasons change it!

I've looked over the diff, and it looks reasonably sane, but I'll let the other builder maintainers comment more specifically -- I just wanted to give you a big :+1:.

File: builder/parser/line_parsers.go [hunk 1]
```diff
@@ -235,7 +246,9 @@ func parseJSON(rest string) (*Node, map[string]bool, error) {
 // so, passes to parseJSON; if not, quotes the result and returns a single
 // node.
 func parseMaybeJSON(rest string) (*Node, map[string]bool, error) {
-	rest = strings.TrimSpace(rest)
```
Comment (tiborvass) [inline builder/parser/line_parsers.go]: @duglin out of curiosity why did you remove the TrimSpace?

File: builder/parser/testfiles-negative/html-page-yes-really-thanks-lk4d4/Dockerfile [hunk 2]
```diff
@@ -1,2 +0,0 @@
-<html>
```
Comment (tiborvass) [inline builder/parser/testfiles-negative/html-page-yes-really-thanks-lk4d4/Dockerfile]: why are you removing this one? is `<html>` a potential instruction? :)

File: builder/parser/line_parsers.go (hunk 1 above)
Comment (tiborvass) [inline builder/parser/line_parsers.go]: oh you moved it to utils.go!

Comment (duglin) [inline builder/parser/line_parsers.go]: @tiborvass because I realized that "rest" (which are the args of the cmd) went thru splitCommand() which already trimmed the spaces.

File: builder/parser/testfiles-negative/html-page-yes-really-thanks-lk4d4/Dockerfile (hunk 2 above)
Comment (duglin) [inline builder/parser/testfiles-negative/html-page-yes-really-thanks-lk4d4/Dockerfile]: This wasn't a failure test due to the invalid command, it was failing because it didn't have any args - which this PR will now allow.  So from a "parser" perspective this test is ok and therefore should be removed as a "negative" test.  Instead, we'll catch the invalid command part of this test later on in the builder, not the parser - and we already have "invalid command" tests in the intergration-cli test suite.

File: integration-cli/docker_cli_build_test.go [hunk 3]
```diff
@@ -4694,3 +4694,51 @@ func TestBuildDockerfileOutsideContext(t *testing.T) {
 
 	logDone("build - Dockerfile outside context")
 }
+
+func TestBuildMissingArgs(t *testing.T) {
+	// test to make sure these cmds w/o any args will generate an error
+	// Notice some commands are missing because its ok for them to
+	// not have any args - like: CMD, RUN, ENTRYPOINT
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin I'm pretty sure all commands need an argument. Even your main PR description only mentions *future* instructions.

Also should we keep in the parser code a map like: `var CmdNArg = map[string]int { "ADD": 1, ... }` or `map[int][]string { 1: {"ADD", "COPY", "ENV", ...}, }` ?

Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: Also wonder whether this should be a unit test instead

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: Actually, all commands don't need an arg today.  I came across an interesting inconsistency.  Today we ban:
```
RUN
```
but allow:
```
RUN []
```
which both end up with the same net result in the resulting image.  The only reason `RUN []` is allowed is because the parser thinks `[]` is an arg, but in reality it eventually gets turned into no args for the `RUN` command. This inconsistency was there for RUN, CMD and ENTRYPOINT - that's why I removed them from the list.

As for making the parser aware of how many args a command needs.... no, I'm not in favor of that.  I really don't like that the parser knows about commands at all. IMO it should be a generic line/word parser and leave understanding the words up to the higher-level code - meaning the builder.  I know we can't get rid of the parser knowing about the commands today (mainly because we have specialized parsers per command) but I don't want to make the situation worse. 
So, I'd prefer not to make this a unit test.

Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin 2 - 0 @tiborvass :)

I just wanna make sure we don't forget to update this test when we add new commands that's all. So maybe a comment in the dispatcher or somewhere else in the builder?

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: that gives me an idea, just a sec....  :-)

**Comment by duglin** [comment]
@tiborvass I made "evaluateTable" in builder public so that we can get the list of Dockerfile commands from the testcase.  So now we won't have to "remember" to update that test - it'll automatically fail if a new Dockerfile command is added and we don't add the "no args" check that the testcase is looking for.
I pushed it as a separate commit so you can see it in isolation - and its easier to backout if we want.
Please take a look and see what you think

**Comment by duglin** [comment]
Not sure why jenkins isn't happy - I don't see that conflict on my machine.
@jfrazelle did I do something wrong?

**Comment by tiborvass** [comment]
@duglin you just need to rebase :)

**Comment by duglin** [comment]
@tiborvass yep, rebased.  You ok with the idea of the testcase reaching back into the builder to get the list of commands?

**Comment by tiborvass** [comment]
@duglin I had the same worry initially. Maybe if we could come up with a small package to put that map in that'd be better. Not sure though.

**Comment by duglin** [comment]
Not sure its worth moving that to a new package. Feels like a lot of possible work/confusion when all we really need to do is expose the list.

File: integration-cli/docker_cli_build_test.go [hunk 4]
```diff
@@ -4822,3 +4823,52 @@ func TestBuildVolumeFileExistsinContainer(t *testing.T) {
...
+		ctx, err := fakeContext(dockerfile, map[string]string{})
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer ctx.Close()
+		var out string
+		if out, err = buildImageFromContext("args", ctx, true); err == nil {
+			t.Fatalf("%s was supposed to fail. Out:%s", cmd, out)
+		}
+		if !strings.Contains(err.Error(), cmd+" requires") {
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: @duglin shouldn't this be `out` instead of `err.Error()` ?

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: the call to buildImageFromContext() is supposed to fail and in those cases `err` contains the output from the build.
See: https://github.com/docker/docker/blob/master/integration-cli/docker_utils.go#L715

Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: My bad, thanks!

File: integration-cli/docker_cli_build_test.go [hunk 5]
```diff
@@ -4822,3 +4823,52 @@ func TestBuildVolumeFileExistsinContainer(t *testing.T) {
...
+			t.Fatalf("%s returned the wrong type of error:%s", cmd, err)
+		}
+
+		ctx.Close()
```
Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: you already have a defer to do that, no?

Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: or i guess we should remove the defer?

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: Not 100% sure of the correct 'go' way to do this, but the `ctx.Close()` here is meant to close the ctx down after each command, as a clean-up.  The defer is there in case we exit prematurely before we get to it. So, this one will negate the need for the defer but is there a way to undo the defer?  Or would it be ok to write it as:
```
out, err = buildImageFromContext("args", ctx, true)
ctx.Close()
if err == nil {
    t.Fatalf("%s was supposed to fail. Out:%s", cmd, out)
}
```
and assume that buildImageFromContext() won't ever panic?  Didn't think that was safe.  So, while (in the happy path cases) the defer is a noop it seemed safer.  Or would you suggest I remove the clean-up one and just do it all in a defer?  Felt odd to have 39 (or so) defers built up.

Comment (tiborvass) [inline integration-cli/docker_cli_build_test.go]: Yeah let's do that instead.

Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: ok done

**Comment by tiborvass** [comment]
For future reference, this PR changes Dockerfile behavior: `RUN`, `ENTRYPOINT`, and `CMD`, are now allowed to have no argument, since it is the same thing as passing `[]` as an argument to any of those instructions.

(waiting for jenkins)

**Comment by duglin** [comment]
any more comments on this one? I'm kind of anxious to resolve this one as its holding up other pieces of work for us.
@jfrazelle @crosbymichael   as @tiborvass pointed out, there is a slight change in Docker's behavior but its actually being more forgiving so not a breaking change, unless people count on getting an error for things like `RUN

**Comment by tiborvass** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Comment by SvenDowideit** [comment]
can someone please explain to me why an empty `RUN`  is desirable?

`CMD` or `ENTRYPOINT` i presume would allow the user to unset an image's CMD/ENRYPOINT - which it would be good to document, so someone wondering how to do it has something to trigger the idea.

**Comment by duglin** [comment]
To me an empty `RUN` isn't necessarily desirable, I actually find it kind of weird and wouldn't mind banning it. However, we allow `RUN []` and some people might have that for some reason today and I won't want to break them.  So, as a side effect of this PR we're making the json and non-json formats equivalent.  ie.  `RUN` == `RUN []` - and now both are legal.

**Comment by crosbymichael** [comment]
LGTM

**Comment by duglin** [comment]
thanks!

**Comment by tiborvass** [comment]
@SvenDowideit some people use `ENTRYPOINT []` to override the parent image.
